interop-java/jvm/src/test/scala/scalaz/zio/interop/javaconcurrent.scala

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2017-2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package scalaz.zio.interop
+
+import java.util.concurrent.{ CompletableFuture, CompletionException, CompletionStage, Future }
+
+import scalaz.zio._
+import scalaz.zio.blocking.{ blocking, Blocking }
+
+import scala.concurrent.ExecutionException
+
+object javaconcurrent {
+
+  implicit class IOObjJavaconcurrentOps(private val ioObj: IO.type) extends AnyVal {
+
+    private def unsafeCompletionStageToIO[A](cs: CompletionStage[A]): Task[A] =
+      IO.effectAsync { cb =>
+        val _ = cs.handle[Unit] { (v: A, t: Throwable) =>
+          if (v != null) {
+            cb(IO.succeed(v))
+          } else {
+            val io = t match {
+              case e: CompletionException =>
+                IO.fail(e.getCause)
+              case t: Throwable =>
+                IO.fail(t)
+            }
+            cb(io)
+          }
+        }
+      }
+
+    def fromCompletionStage[A, E >: Throwable](csIo: IO[E, CompletionStage[A]]): IO[E, A] =
+      csIo.flatMap(unsafeCompletionStageToIO)
+
+    def fromCompletionStage[A](cs: CompletionStage[A]): Task[A] =
+      IO.suspend {
+        unsafeCompletionStageToIO(cs)
+      }
+
+    private def unsafeFutureJavaToIO[A](future: Future[A]): Task[A] = {
+      def unwrap[B](f: Future[B]): Task[B] =
+        IO.flatten {
+          IO.effectTotal {
+            try {
+              val result = f.get()
+              IO.succeed(result)
+            } catch {
+              case e: ExecutionException =>
+                IO.fail(e.getCause)
+              case _: InterruptedException =>
+                IO.interrupt
```

Comment:
In this case, you have to clear the thread interrupt status, with `Thread.interrupted`. Just call that right before returning `IO.interrupt`.